src/device/linux-arm-rpi2/gpioctl_linux-arm-rpi2.cpp

```diff
@@ -0,0 +1,295 @@
+
+int GpioControlInst::open_device(const char* devicefile, bool wmode) {
+  int fd = -1;
+  int retry = GPIO_OPEN_RETRY;
+  int sleepvalue = GPIO_RETRY_WAIT;
+  int flags = wmode ? O_WRONLY : O_RDONLY;
+
+  // creating virtual file for device control takes some time
+  while (fd < 0 && retry > 0) {
+    fd = open(devicefile, flags);
+    if (fd < 0) {
+      usleep(sleepvalue);
+      // try to wait more longer
+      sleepvalue += GPIO_RETRY_WAIT;
+      // enable port(export), which creates sym-link, takes about 45msec
+    }
+    retry--;
+  }
+  return fd;
+}
+
+
+int GpioControlInst::enable_port(uint8_t gpionum, bool enable) {
+  int fd = open_device(enable ? gpio_export_1 : gpio_export_0, true);
+  if (fd < 0) {
+    DLOG("RPi2 GPIO, failed to set gpio %d", gpionum);
+    return IOTJS_GPIO_WRITEFAIL;
+  }
+  ssize_t towrite = snprintf(gpio_file_buff, GPIO_RPI2_MAX_PATH, "%d", gpionum);
+  // if gpio file exist(result of previous action) write result can be -1
+  // so don't check written value
+  ssize_t written = write(fd, gpio_file_buff, towrite);
+  close(fd);
+  (void)written; // write() requres return value or compiler warns.
+
+  return 0;
+}
+
+
+int GpioControlInst::PinMode(uint32_t portpin) {
+  if (_fd <= 0)
+    return IOTJS_GPIO_NOTINITED;
+
+  uint8_t gpionum = rpi2_get_gpio(portpin & 0xff);
+  int ret = 0;
+
+  if (portpin & GPIO_RPI2_MODE_ENABLE) {
+    ret = enable_port(gpionum, true);
+      // if port enable failes, setting futhur mode may not be possible
+    if (ret < 0)
+      return ret;
+
+    static const char* dir_out = "out";
+    static const char* dir_inp = "in";
+    const char* direction = NULL;
+    int dirleng = 0;
+
+    if (portpin & GPIO_RPI2_MODE_OUTPUT) {
+      direction = dir_out;
+      dirleng = 3;
+    }
+    else {
+      direction = dir_inp;
+      dirleng = 2;
+    }
+    DDDLOG("RPi2 GPIO direction %d : %s\n", gpionum, direction);
+
+    snprintf(gpio_file_buff, GPIO_RPI2_MAX_PATH, gpio_dirfmt, gpionum);
+    int fd = open_device(gpio_file_buff, true);
+    if (fd < 0) {
+      DLOG("RPi2 GPIO, failed to set direction %d", gpionum);
+      return -1;
+    }
+    if (write(fd, direction, dirleng) < 0) {
+      DLOG("RPi2 GPIO, failed to set direction %d", gpionum);
+      ret = -1;
+    }
+    close(fd);
+  }
+  else {
+    if (enable_port(gpionum, false) < 0)
+      ret = -1;
+  }
+
+  return ret;
+}
+
+
+int GpioControlInst::WritePin(uint32_t portpin, uint8_t data) {
+  if (_fd <= 0)
+    return IOTJS_GPIO_NOTINITED;
+
+  static const char* val_0 = "0";
+  static const char* val_1 = "1";
+  uint8_t gpionum = rpi2_get_gpio(portpin & 0xff);
+
+  snprintf(gpio_file_buff, GPIO_RPI2_MAX_PATH, gpio_valfmt, gpionum);
+  int fd = open_device(gpio_file_buff, true);
+  if (fd < 0) {
+    DLOG("RPi2 GPIO, failed to write %d", gpionum);
+    return IOTJS_GPIO_WRITEFAIL;
+  }
+
+  int ret = 0;
+  ssize_t written = write(fd, data ? val_1 : val_0, 1);
```

Comment:
Why do we have to write only "0" or "1" instead of given data.
Are there problems with `write(fd, &data, sizeof(data));`?